### PR TITLE
crimson/osd/backfill_state: do at least one time of replica scanning if necessary in the Enqueuing state

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Fixes: https://tracker.ceph.com/issues/68175
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by xxhdx1985126** [comment]
jenkins test api

**Comment by xxhdx1985126** [comment]
jenkins test make check

File: src/crimson/osd/backfill_state.cc [hunk 1]
```diff
@@ -225,7 +225,7 @@ bool BackfillState::Enqueuing::should_rescan_primary(
   const BackfillInterval& backfill_info) const
 {
   return backfill_info.begin <= earliest_peer_backfill(peer_backfill_info) &&
-	 !backfill_info.extends_to_end();
+	 !backfill_info.extends_to_end() && backfill_info.empty();
```
Comment (athanatos) [inline src/crimson/osd/backfill_state.cc:228]: I'm confused by this condition and the one  in ReplicasScanning::replica_needs_scan.

```backfill_info.begin <= earliest_peer_backfill(peer_backfill_info)``` and the corresponding check in replica_needs_scan ```replica_backfill_info.begin <= local_backfill_info.begin``` seem wrong.  It seems like the check should be whether there's overlap between the interval and the next thing that needs to be backfilled -- something like

```backfill_info.end <= earliest_peer_backfill(peer_backfill_info)``` and ```replica_backfill_info.end <= local_backfill_info.begin```

The empty() check is probably fine, but I'm not sure if it's actually necessary.  For there to be no overlap, the contents of the previous scan would have to already have been backfilled and removed.

File: src/crimson/osd/backfill_state.cc [hunk 2]
```diff
@@ -327,15 +327,24 @@ BackfillState::Enqueuing::Enqueuing(my_context ctx)
...
+    // the queueing.
+    logger().debug("{}: reached end for current local chunk", __func__);
+    post_event(RequestPrimaryScanning{});
+    return;
+  } else if (should_rescan_replicas(backfill_state().peer_backfill_info,
+				    primary_bi)) {
+    // Count simultaneous scans as a single op and let those complete
+    post_event(RequestReplicasScanning{});
+    return;
+  }
```
Comment (athanatos) [inline src/crimson/osd/backfill_state.cc:342]: This works because we now always loop back to Enqueuing, right?  Seems cleaner.

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/crimson/osd/backfill_state.cc (hunk 2 above)
Comment (xxhdx1985126) [reply src/crimson/osd/backfill_state.cc:342]: Yes, that's right.

File: src/crimson/osd/backfill_state.cc (hunk 1 above)
Comment (xxhdx1985126) [reply src/crimson/osd/backfill_state.cc:228]: > backfill_info.begin <= earliest_peer_backfill(peer_backfill_info) and the corresponding check in replica_needs_scan replica_backfill_info.begin <= local_backfill_info.begin seem wrong. It seems like the check should be whether there's overlap between the interval and the next thing that needs to be backfilled -- something like

I think those checks are copied from `PrimaryLogPG::recover_backfill()`

https://github.com/ceph/ceph/blob/d7a74a60764825edeb5e7d99435737917447fd1a/src/osd/PrimaryLogPG.cc#L13946-L13947
https://github.com/ceph/ceph/blob/d7a74a60764825edeb5e7d99435737917447fd1a/src/osd/PrimaryLogPG.cc#L13965-L13966

Comment (athanatos) [reply src/crimson/osd/backfill_state.cc:228]: Ah, I see now.  If ```backfill_info.begin <= earliest_peer_backfill(peer_backfill_info)``` and it's empty (and doesn't extend to end), they don't actually overlap and we need to rescan primary to get more entries.  Checking end could also work, but this change to match PrimaryLogPG is fine.

**Review by athanatos — APPROVED** [review]

**Comment by xxhdx1985126** [comment]
jenkins test make check

**Comment by xxhdx1985126** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-10-13_10:25:17-crimson-rados-wip-matanb-crimson-testing-oct-10-distro-crimson-smithi/

Failures are 2 infra and one seastore

**Pull request merged**